pymc3/distributions/timeseries.py

```diff
@@ -307,10 +309,15 @@ def _random(self, sigma, mu, size, sample_shape):
         if size[len(sample_shape)] == sample_shape:
             axis = len(sample_shape)
         else:
-            axis = 0
+            axis = len(size) - 1
         rv = stats.norm(mu, sigma)
         data = rv.rvs(size).cumsum(axis=axis)
-        data = data - data[0]  # TODO: this should be a draw from `init`, if available
+        data = np.array(data)
+        if len(data.shape)>1:
+            for i in range(data.shape[0]):
+                data[i] = data[i] - data[i][0]
```

Comment:
This assumed that the second axis of the `data` will be the time dimension. This isn't true in general, for example consider the case `rv = GaussianRandomWalk(..., shape=(10, 2))`.
- A single draw from this distribution will have shape (10, 2).
- The distribution's `logp` implies that the dimension with 10 entries is the "time" axis. This means `rv.shape[-2] == 10` and it will be the time axis along which the `cumsum` has to be applied.
- `rv.random().shape == (10, 2)` but your code will assume that the `cumsum` should be applied to the last axis.
- `rv.random(size=(3, 4)).shape == (3, 4, 10, 2)` but your code will assume that the time axis is axis=1, when it should really be axis=2.

Everything turns into an even bigger mess when you make some of the distribution's parameters be random variables as well. In those cases, you'll run into broadcasting errors between the GRW's time axis and some of its parameters sample_shape axis.

Reply:
@lucianopaz Do you have any outline of the cases that I have to keep track of while I code for the axis selection?
Also, can you please refer me to any material that explains the interplay between size and shape in pymc3. I read your blog on shape handling but could not understand how it interacts with size.

Reply:
I'll be able to write down a list of tests later today. About the resources to read, I'm afraid that all we have are the [shape_utils docs](https://docs.pymc.io/api/shape_utils.html) and the thing that I wrote in my blog.

The relation between the `size` parameter you pass to `random` and the output's `shape` is the following. Imagine you created an RV like this: `x = GaussianRandomWalk(..., shape=(3, 4))`. The distribution's shape is something I call the core shape. When you call `x.random(size=(1, 2))`, the `size` is something that we are starting to call the `sample_shape` (like what tensorflow probability does). The final outputs shape should always be equal to `sample_shape + core_shape`, which in this case is `(1, 2) + (3, 4)`. So the resulting array will have `shape = (1, 2, 3, 4)`.

Reply:
As I promised earlier, the list of tasks that I think should be tested to verify that GRW works are to construct the distribution with different types and shapes of parameters `mu` and different `shape` values (what I called the core shape in a previous comment). Then you should call `random` on it with different `size` arguments and verify that the result has the correct shape (ideally you should test `size=None`, and then a scalar size, like `size=7`, and a 2d size like `size=(5, 7)`).

The combinations of parameters and shape values that I suggest you test are:
1. `mu=0`, `shape=(4,)`
2. `mu=np.arange(4)`, `shape=(4,)`
3. `mu=np.arange(4)`, `shape=(3, 4,)`
4. `mu=Normal(0, 1)`, `shape=(4,)`
5. `mu=Normal(np.arange(4), 1e-4, shape=4)`, `shape=(3, 4,)`

You can write a parameterized test that performs all 15 combinations. Once you get the shape stuff right, the `axis` won't be so hard to set correctly.

Reply:
```
with pymc3.Model() as pmodel1:
    grw1 = pymc3.GaussianRandomWalk('grw1', mu=numpy.arange(4), shape=(3,4,))

grw1.random(size = None).shape
```

When `shape `is introduced in the form of` (3,4,)`, I am getting an error message 
`ValueError: Input dimension mis-match. (input[0].shape[1] = 4, input[1].shape[1] = 3)`
for the line     

`grw1 = pymc3.GaussianRandomWalk('grw1', mu=numpy.arange(4), shape=(3,4,))
`

It does not seem to me an issue of the `_random` method. Can you please advise me on this issue?

PFB the detailed traceback:

```
ValueError                                Traceback (most recent call last)
<ipython-input-24-177b01fab8f8> in <module>
      1 with pymc3.Model() as pmodel1:
----> 2     grw1 = pymc3.GaussianRandomWalk('grw1', mu=pymc3.Normal('mu',numpy.arange(4),1e-4,shape = 4), shape=(3,4,))
      3 
      4 grw1.random(size = None).shape

/mnt/c/Users/RISHAV/pymc3/pymc3/distributions/distribution.py in __new__(cls, name, *args, **kwargs)
     81         else:
     82             dist = cls.dist(*args, **kwargs)
---> 83         return model.Var(name, dist, data, total_size, dims=dims)
     84 
     85     def __getnewargs__(self):

/mnt/c/Users/RISHAV/pymc3/pymc3/model.py in Var(self, name, dist, data, total_size, dims)
   1070                 with self:
   1071                     var = FreeRV(
-> 1072                         name=name, distribution=dist, total_size=total_size, model=self
   1073                     )
   1074                 self.free_RVs.append(var)

/mnt/c/Users/RISHAV/pymc3/pymc3/model.py in __init__(self, type, owner, index, name, distribution, total_size, model)
   1591                 np.ones(distribution.shape, distribution.dtype) * distribution.default()
   1592             )
-> 1593             self.logp_elemwiset = distribution.logp(self)
   1594             # The logp might need scaling in minibatches.
   1595             # This is done in `Factor`.

/mnt/c/Users/RISHAV/pymc3/pymc3/distributions/timeseries.py in logp(self, x)
    272             x_i = x[1:]
    273             mu, sigma = self._mu_and_sigma(self.mu, self.sigma)
--> 274             innov_like = Normal.dist(mu=x_im1 + mu, sigma=sigma).logp(x_i)
    275             return self.init.logp(x[0]) + tt.sum(innov_like)
    276         return self.init.logp(x)

/mnt/c/Users/RISHAV/pymc3/myvenv/lib/python3.6/site-packages/theano/tensor/var.py in __add__(self, other)
    126     def __add__(self, other):
    127         try:
--> 128             return theano.tensor.basic.add(self, other)
    129         # We should catch the minimum number of exception here.
    130         # Otherwise this will convert error when Theano flags

/mnt/c/Users/RISHAV/pymc3/myvenv/lib/python3.6/site-packages/theano/gof/op.py in __call__(self, *inputs, **kwargs)
    672                 thunk.outputs = [storage_map[v] for v in node.outputs]
    673 
--> 674                 required = thunk()
    675                 assert not required  # We provided all inputs
    676 

/mnt/c/Users/RISHAV/pymc3/myvenv/lib/python3.6/site-packages/theano/gof/op.py in rval()
    860 
    861         def rval():
--> 862             thunk()
    863             for o in node.outputs:
    864                 compute_map[o][0] = True

/mnt/c/Users/RISHAV/pymc3/myvenv/lib/python3.6/site-packages/theano/gof/cc.py in __call__(self)
   1737                 print(self.error_storage, file=sys.stderr)
   1738                 raise
-> 1739             reraise(exc_type, exc_value, exc_trace)
   1740 
   1741 

/mnt/c/Users/RISHAV/pymc3/myvenv/lib/python3.6/site-packages/six.py in reraise(tp, value, tb)
    701             if value.__traceback__ is not tb:
    702                 raise value.with_traceback(tb)
--> 703             raise value
    704         finally:
    705             value = None

ValueError: Input dimension mis-match. (input[0].shape[1] = 4, input[1].shape[1] = 3)

```

Reply:
Oh! You're right that this has nothing to do with `random`. It's great that the new tests unearthed this separate issue. It looks like `logp` made some incorrect assumptions on the shape of `mu` and `sigma`. We can also fix that issue here, but maybe we can try to simplify things by placing the time axis in the last dimension instead of the first one. I'll get back to you later.

Reply:
Should I raise a new issue for this?  I think the current PR has solved the issue of `sample_prior_predictive` for usual cases so it should be used as a temporary fix.

Reply:
Yes, that would be best.